### PR TITLE
[ESIMD] Implement property-based gather(usm, ...)

### PR DESCRIPTION
**Pull request by v-klochkov**
This patch also supports gather and masked-gather of any length N if it does not use L1/L2 hints of VS>1.

Additionally for gathers without L1/L2 vs VS>1 this patch replaces the calls of GenX SVM gather calls with LLVM IR
if the macro __ESIMD_GATHER_SCATTER_LLVM_IR is defined by user. If it not defined, then using masked gathers
with pass_thru operand requires DG2/PVC.

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 1]
```diff
@@ -67,65 +67,537 @@ __ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc) {
...
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   simd_mask<N / VS> mask, simd<T, N> pass_thru,
+///                   PropertyListT props = {});                   // (usm-ga-1)
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   simd_mask<N / VS> mask,
+///                   PropertyListT props = {});                   // (usm-ga-2)
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   PropertyListT props = {});                   // (usm-ga-3)
+///
+/// The next 3 functions are similar to the above and added for convenience.
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:208]:
```suggestion
/// The next 3 functions are similar to the above and were added for convenience.
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 2]
```diff
@@ -67,65 +67,537 @@ __ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc) {
...
+/// template <typename T, int N, int VS, typename OffsetT,
+///           typename PropertyListT = empty_properties_t>
+/// simd<T, N> gather(const T *p, simd<OffsetT, N / VS> byte_offsets,
+///                   simd_mask<N / VS> mask, simd<T, N> pass_thru,
+///                   PropertyListT props = {});                   // (usm-ga-1)
+#ifndef __ESIMD_GATHER_SCATTER_LLVM_IR
+/// Supported platforms: DG2, PVC only - Temporary restriction for the variant
+/// with pass_thru operand.
+#endif // __ESIMD_GATHER_SCATTER_LLVM_IR
+/// Loads ("gathers") elements ov the type 'T' from memory locations addressed
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:246]:
```suggestion
/// Loads ("gathers") elements of the type 'T' from memory locations addressed
```

same thing in a few other spots

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 3]
```diff
@@ -67,65 +67,537 @@ __ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc) {
...
+/// @return A vector of elements read.
+template <typename T, int N, int VS = 1, typename OffsetObjT,
+          typename OffsetRegionT,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    ext::oneapi::experimental::is_property_list_v<PropertyListT>, simd<T, N>>
+gather(const T *p, simd_view<OffsetObjT, OffsetRegionT> byte_offsets,
+       simd_mask<N / VS> mask, simd<T, N> pass_thru, PropertyListT props = {}) {
+  static_assert(std::is_integral_v<typename OffsetRegionT::element_type>,
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:536]: Shouldn't the existing assert in the called function that uses the `read` result catch the bad case already?

File: sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp [hunk 4]
```diff
@@ -54,7 +54,7 @@ template <typename T, int N> bool test(queue &Q) {
   T *Dst = Ptr2.get();
 
   for (int I = 0; I < N; ++I) {
-    Src[I] = I + 1;
+    Src[I] = I + 1 + 0.5;
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp:57]: Sorry can you explain this change?

File: sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp [hunk 5]
```diff
@@ -123,6 +123,7 @@ int main(void) {
   Pass &= test<int32_t, 16>(Q);
   Pass &= test<int32_t, 32>(Q);
 #endif
+
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp:126]: Huge nitpick: Remove whitespace changes in this file

File: sycl/test-e2e/ESIMD/unified_memory_api/gather_usm_pvc.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
+//==------- gather_usm_pvc.cpp - DPC++ ESIMD on-device test ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu-intel-pvc
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/unified_memory_api/gather_usm_pvc.cpp:8]: Is there any DG2 functionality we can test?

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 7]
```diff
@@ -67,65 +67,537 @@ __ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc) {
...
+  static_assert(Alignment >= sizeof(T),
+                "gather() requires at least element-size alignment");
+  constexpr auto L1Hint =
+      detail::getPropertyValue<PropertyListT, cache_hint_L1_key>(
+          cache_hint::none);
+  constexpr auto L2Hint =
+      detail::getPropertyValue<PropertyListT, cache_hint_L2_key>(
+          cache_hint::none);
+
+  // Use LSC lowering if L1/L2 or VS > 1 requires that. Also, if masked
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:289]:
```suggestion
  // Use LSC lowering if L1/L2 or VS > 1. Also, if masked
```

Comment a bit unclear to me. Maybe this?

File: sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp (hunk 4 above)
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp:57]: I did not like the fact the data is basically int. This change is the attempt to make the test a bit more strict and catch the cases where our API implementation can use fp-to-int vector conversion where fp-to-int and back bit-cast was needed.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 8]
```diff
@@ -67,65 +67,537 @@ __ESIMD_API SurfaceIndex get_surface_index(AccessorTy acc) {
...
+///           typename PropertyListT = empty_properties_t>
+/// simd<T, N> gather(const T *p, simd<OffsetT, N> byte_offsets,
+///                   simd_mask<N> mask, simd<T, N> pass_thru,
+///                   PropertyListT props = {});                   // (usm-ga-4)
+/// simd<T, N> gather(const T *p, simd<OffsetT, N> byte_offsets,
+///                   simd_mask<N> mask, PropertyListT props = {});// (usm-ga-5)
+/// simd<T, N> gather(const T *p, simd<OffsetT, N> byte_offsets,
+///                   PropertyListT props = {});                   // (usm-ga-6)
+///
+/// The next 5 functions are variations of the first 3 above (usm-ga-1,2,3)
```
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: typo. fixed in f541a69d936f5c31befd154497f8f1e9f1af7b57

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 1 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed in f541a69d936f5c31befd154497f8f1e9f1af7b57

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 2 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed in f541a69d936f5c31befd154497f8f1e9f1af7b57

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 3 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Yes, I removed this assert here. Fixed in f541a69d936f5c31befd154497f8f1e9f1af7b57

File: sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp (hunk 5 above)
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/api/svm_gather_scatter.cpp]: Ok, removed in f541a69d936f5c31befd154497f8f1e9f1af7b57

File: sycl/test-e2e/ESIMD/unified_memory_api/gather_usm_pvc.cpp (hunk 6 above)
Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/unified_memory_api/gather_usm_pvc.cpp]: gather is supported similarly for DG2 and PVC. I enabled the test for DG2. Fixed in f541a69d936f5c31befd154497f8f1e9f1af7b57

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 7 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp]: Fixed. Thank you. f541a69d936f5c31befd154497f8f1e9f1af7b57

**Review by sarnex — APPROVED** [review]

**Pull request merged**